walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,85 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  private static int DAY_LENGTH = 24*60;
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    Set<String> attendees =  new HashSet<String>();
+    int meetingDur = (int)request.getDuration();
+    attendees.addAll(request.getAttendees());
+    if (meetingDur > DAY_LENGTH){
+      return possibleTimes;
+    }
+    if (events.isEmpty() || attendees.isEmpty()){
+      possibleTimes.add(TimeRange.WHOLE_DAY);
+      return possibleTimes;
+    }
+    /*A time is considered to be busy when an attendee is in an event & the meeting request*/
+    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    for(Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      Set<String> overlap = new HashSet<>(eventAttendees);
+      overlap.retainAll(attendees);
+      if (!overlap.isEmpty()){
+        busyTimes.add(event.getWhen());
+      }
+    }
+  /*Combines overlapping busy times into an ArrayList*/
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    TimeRange[] blockedTimes = ((List<TimeRange>) busyTimes).toArray(new TimeRange[busyTimes.size()]); 
+    ArrayList<TimeRange> busyWithOverlap = new ArrayList<TimeRange>();
+    if(blockedTimes.length > 0) {
+      busyWithOverlap.add(blockedTimes[0]);
+    }
+    int index = 0;
+    for (int i=0; i < blockedTimes.length; i++){
+      if (blockedTimes[i].overlaps(busyWithOverlap.get(index))){
+        int meetingStart = Math.min(busyWithOverlap.get(index).start(), blockedTimes[i].start());
+        int meetingEnd = Math.max(busyWithOverlap.get(index).end(), blockedTimes[i].end()) - 1;
+        busyWithOverlap.set(index, TimeRange.fromStartEnd(meetingStart, meetingEnd, true));
+      }
+      else {
+        busyWithOverlap.add(blockedTimes[i]);
+        index ++;
+      }
+    }
```

Comment:
It looks like this could be a good chunk for a helper method!

Reply:
Good idea! Put it in the last PR, it definitely helps readability